[flang] Add EXECUTE_COMMAND_LINE runtime and lowering intrinsics implementation

yiwu0b11 opened the pull request:
This patch add support of intrinsics Fortran 2008 EXECUTE_COMMAND_LINE.
The patch contains both the lowering and the runtime code and works on
both Windows and Linux. The patch contains a list of commits, to convey
the authorship and the history of changes. Some implementation specifics
or status has been added to `flang/docs/Intrinsics.md`.

I have provided a summary of the usage and the options required for the
`EXECUTE_COMMAND_LINE intrinsic`. The intrinsic supports both a synchronous
(by default) and an asynchronous option.

| System  | Mode  | Implemention              |
|---------|-------|---------------------------|
| Linux   | Sync  | std::system()             |
| Windows | Sync  | std::system()             |
| Linux   | Async | fork()  |
| Windows | Async | CreateProcess             |

Support for the SYSTEM GNU extension will be added in a separate PR.

Co-authored with @jeffhammond

jeffhammond committed:
add execute_command_line

Signed-off-by: Jeff Hammond <jeff.science@gmail.com>

jeffhammond committed:
this is definitely not completely correct

jeffhammond committed:
definitely not right either

jeffhammond committed:
definitely not right either

jeffhammond committed:
more correct

jeffhammond committed:
more correct

jeffhammond committed:
remove trailing space

jeffhammond committed:
5 args not 6

jeffhammond committed:
i have no idea what i am doing

jeffhammond committed:
runtime part of this

jeffhammond committed:
WIP

jeffhammond committed:
where is the pattern for IsValidLogicalDescriptor

yiwu0b11 committed:
Merge remote-tracking branch 'jeffhammond/jeffhammond-execute_command_line' into execute_command_line

yiwu0b11 committed:
some small fixes to make it build

yiwu0b11 committed:
Add simple support of async and sync execution, might not work on Windows

yiwu0b11 committed:
some changes to match the standard 16.9.73

exitstat: If sync, assigned processor-dependent exit status. Otherwise unchanged.
cmdstast: Assigned 0 as specifed in standard, if error then overwrite.
If a condition occurs that would assign a nonzero value to CMDSTAT but
the CMDSTAT variable is not present, error termination is initiated.

yiwu0b11 committed:
Fixing error termination Linux only

Work and test on Linux, both sync and async mode.
Sync mode: termination will terminate directly
Async mode: will only terminate the child/async process, no effect on parent
Standard: If a condition occurs that would assign a nonzero value to CMDSTAT,
but the CMDSTAT variable is not present, error termination is initiated.

Commit:
Added EXECL_ERR if exit==-1, update docs

yiwu0b11 committed:
Add runtime tests

yiwu0b11 committed:
Add lowering intrinsic tests

yiwu0b11 committed:
add more docs, format, and brace initialization

Commit:
add test fixes on Windows and remove `address_of` in testing

yi-wu-arm commented:
Implementation specifics or status on `flang/docs/Intrinsics.md`

### Standard Intrinsics: EXECUTE_COMMAND_LINE

#### Usage and Info

- **Standard:** Fortran 2008 and later, specified in 16.9.73
- **Class:** Subroutine
- **Syntax:** `CALL EXECUTE_COMMAND_LINE(COMMAND [, WAIT, EXITSTAT, CMDSTAT, CMDMSG ])`
- **Arguments:**

  | Argument  | Description                                                  |
  |-----------|--------------------------------------------------------------|
  | `COMMAND` | Shall be a default CHARACTER scalar.                         |
  | `WAIT`    | (Optional) Shall be a default LOGICAL scalar.                |
  | `EXITSTAT`| (Optional) Shall be an INTEGER of the default kind.          |
  | `CMDSTAT` | (Optional) Shall be an INTEGER of the default kind.          |
  | `CMDMSG`  | (Optional) Shall be a CHARACTER scalar of the default kind.  |

#### Implementation Specifics

- **`COMMAND`:**
  - Must be preset.

- **`WAIT`:**
  - If set to `false`, the command is executed asynchronously. If not preset or set to `false`, it is executed synchronously.
  - Sync: achieved by passing command into `std::system` on all systems.
  - Async: achieved by calling a `fork()` on POSIX-compatible systems, or `CreateProcess()` on Windows.

- **`CMDSTAT`:**
  - -2: No error condition occurs, but `WAIT` is present with the value `false`, and the processor does not support asynchronous execution.
  - -1: The processor does not support command line execution.
  - \+ (positive value): An error condition occurs.
    - 1: Fork Error, where `pid_t < 0`, would only occur on POSIX-compatible systems.
    - 2: Execution Error, a command exits with status -1.
    - 3: Invalid Command Error, determined by the exit code depending on the system.
      - On Windows, if the exit code is 1.
      - On POSIX-compatible systems, if the exit code is 127 or 126.
    - 4: Signal error, either it is stopped or killed by signal, would only occur on POSIX-compatible systems.
  - 0: Otherwise.

- **`CMDMSG`:**
  - If an error condition occurs, it is assigned an explanatory message. Otherwise, it remains unchanged.
  - If a condition occurs that would assign a nonzero value to `CMDSTAT` but the `CMDSTAT` variable is not present, error termination is initiated.
    - On POSIX-compatible systems, this applies to both synchronous and asynchronous error termination. When the execution mode is set to async with error termination, the child process (async process) will be terminated with no effect on the parent process (continues).
    - On Windows, this only applies to synchronous error termination.

klausler commented:
That makes sense, given that you want to use `system()` rather than `fork()` and `execvp()`.  On the asynchronous POSIX path you're actually creating two processes rather than one, unfortunately.

Do you ever use `waitpid()` to harvest these completed processes?  If not, they'll occupy system resources until the program exits.

yi-wu-arm commented:
> That makes sense, given that you want to use `system()` rather than `fork()` and `execvp()`. On the asynchronous POSIX path you're actually creating two processes rather than one, unfortunately.
> 
> Do you ever use `waitpid()` to harvest these completed processes? If not, they'll occupy system resources until the program exits.

I think `system` is internal implemented with `fork(0` and `execl` synchronously, so yes it will create two process not one. On Linux async, there is an `exit(status)` to quit the async process. 
On Windows there is none. Although I have this, when `CreateProcess` return a success status. But I don't think it will actually do anything since I didn't wait the process to finish.  (But if I wait, that will make it sync instead of async...)
```cpp
      CloseHandle(pi.hProcess);
      CloseHandle(pi.hThread);
```

yiwu0b11 committed:
terminated children do not become zombies on Linux

signal(SIGCHLD, SIG_IGN);
https://man7.org/linux/man-pages/man2/sigaction.2.html

POSIX.1-1990 disallowed setting the action for SIGCHLD to
SIG_IGN.  POSIX.1-2001 and later allow this possibility, so that
ignoring SIGCHLD can be used to prevent the creation of zombies
(see wait(2)).  Nevertheless, the historical BSD and System V
behaviors for ignoring SIGCHLD differ, so that the only
completely portable method of ensuring that terminated children
do not become zombies is to catch the SIGCHLD signal and perform
a wait(2) or similar.

yiwu0b11 committed:
required para use reference not pointer, ensure command isn null-terminated

Commit:
fixes on Windows

yiwu0b11 committed:
clang-format

yi-wu-arm commented:
@klausler Any thought of how zombie process is handled?  Thanks in advance.

yiwu0b11 committed:
fix null-terminated if logic and minor fixes

klausler commented:
> @klausler Any thought of how zombie process is handled? Thanks in advance.

`signal(SIGCHLD, SIG_IGN)` sounds good to me.

yi-wu-arm commented:
Hi @jeanPerier Sorry to throw a ping. Do you have any comment about this patch? Thanks in advance.

jeanPerier commented on flang/test/Lower/Intrinsics/execute_command_line-optional.f90:
```diff
@@ -0,0 +1,51 @@
+  LOGICAL, OPTIONAL :: isWait
+  ! Note: command is not optional in execute_command_line and must be present
+  call execute_command_line(command, isWait, exitVal, cmdVal, msg)
+! CHECK:         %0 = fir.declare %[[cmdstatArg]] {fortran_attrs = #fir.var_attrs<optional>, uniq_name = "_QFall_args_optionalEcmdval"} : (!fir.ref<i32>) -> !fir.ref<i32>
```
The test is using hard coded SSA numbers, this is brittle (especially when testing with -emit-fir which runs some high level pass like CSE).

You can use this script to help with the check generation: https://github.com/llvm/llvm-project/blob/main/mlir/utils/generate-test-checks.py

jeanPerier reviewed (commented):
Hi @yi-wu-arm, the lowering code looks good to me. Please just update the tests to not test for specific SSA IR numbers.

I do not have specific comment on the runtime, I am not familiar enough with the corresponding C/C++ APIs.

yiwu0b11 committed:
remove hard-coded SSA number from test

yiwu0b11 committed:
move EnsureNullTerminated to tools.h, free memory after use.

On Linux, malloc is freed in parent, child process will be cleaned up on termination.
Same approach was taken on Windows as well, but does it work in the same way?

yiwu0b11 committed:
move StoreToDescriptor functions to tools.h

yi-wu-arm commented:
On Linux, malloc is freed in parent, child process will be cleaned up on termination.
Windows: same as Linux, when a child process is terminated OS will free the allocated memory.
https://learn.microsoft.com/en-us/windows/win32/procthread/terminating-a-process `Any resources allocated by the process are freed.`

klausler commented on flang/runtime/tools.cpp:
```diff
@@ -173,5 +173,145 @@ RT_API_ATTRS void ShallowCopy(const Descriptor &to, const Descriptor &from) {
+// Returns the length of the \p string. Assumes \p string is valid.
+RT_API_ATTRS std::int64_t StringLength(const char *string) {
+  std::size_t length{std::strlen(string)};
+  if constexpr (sizeof(std::size_t) < sizeof(std::int64_t)) {
```
Why can't this whole function be just `static_cast<std::int64_t>(std::strlen(string))`???

yi-wu-arm replied:
I believe at the time when this code was written, `std::strlen()` is not reliable so they add checks. Done

klausler commented on flang/runtime/tools.cpp:
```diff
@@ -173,5 +173,145 @@ RT_API_ATTRS void ShallowCopy(const Descriptor &to, const Descriptor &from) {
+}
+
+RT_API_ATTRS void FillWithSpaces(const Descriptor &value, std::size_t offset) {
+  if (offset < value.ElementBytes()) {
```
This will work only for a scalar character value of kind 1, so you should check that.

klausler replied:
If it's not a safe general-purpose utility routine, then don't put it into `tools.cpp`.  Or leave it here and add some checking.

yi-wu-arm replied:
agree, moved back to `command.cpp`

klausler commented on flang/runtime/tools.cpp:
```diff
@@ -173,5 +173,145 @@ RT_API_ATTRS void ShallowCopy(const Descriptor &to, const Descriptor &from) {
+      value->rank() == 0;
+}
+
+RT_API_ATTRS bool IsValidIntDescriptor(const Descriptor *length) {
```
What it `length` is a null pointer?

klausler commented on flang/runtime/execute.cpp:
```diff
@@ -0,0 +1,184 @@
+      exit(status);
+    }
+#endif
+    FreeMemory((void *)newCmd);
```
Move this deallocation out one level of braces so that it also handles the synchronous case.

yi-wu-arm replied:
Done

yiwu0b11 committed:
minor fixes in tools and move memory deallocate to bottom.

yiwu0b11 committed:
Merge remote-tracking branch 'yi/main' into yi_dev_flang

yiwu0b11 committed:
clang-format

klausler commented on flang/runtime/tools.h:
```diff
@@ -411,6 +412,48 @@ RT_API_ATTRS void ShallowCopy(const Descriptor &to, const Descriptor &from,
+
+RT_API_ATTRS std::size_t LengthWithoutTrailingSpaces(const Descriptor &d);
+
+// Returns the length of the \p string. Assumes \p string is valid.
```
The comment doesn't specify how the length of the string is defined.  If it is NUL terminated, how is this different from `std::strlen()`?

yi-wu-arm replied:
moved back to `command.cpp` since it is only used there.

klausler commented on flang/runtime/tools.h:
```diff
@@ -411,6 +412,48 @@ RT_API_ATTRS void ShallowCopy(const Descriptor &to, const Descriptor &from,
+    const char *rawValue, std::int64_t rawValueLength, const Descriptor *errmsg,
+    std::size_t offset = 0);
+
+void CopyCharToDescriptor(
```
Does this routine copy a single character, or some kind of NUL terminated string, or something else?

yi-wu-arm replied:
add comment.
```cpp
// Copy a null-terminated character array \p rawValue to descriptor \p value.
// The copy starts at the given \p offset, if not present then start at 0.
// If descriptor `errmsg` is provided, error messages will be stored to it.
// Returns stats specified in standard.
```
the returned stats code is only used in `command.cpp` but functionality of copying is used in both `command.cpp` and `execute.cpp`

klausler commented on flang/runtime/tools.h:
```diff
@@ -411,6 +412,48 @@ RT_API_ATTRS void ShallowCopy(const Descriptor &to, const Descriptor &from,
+
+template <int KIND> struct FitsInIntegerKind;
+
+RT_API_ATTRS bool FitsInDescriptor(
```
This routine looks like a Boolean predicate but why does it require a `Terminator`?  Can it crash?

yi-wu-arm replied:
it calls `ApplyIntegerKind` which take a `bool` as an input, but it's not used in somewhere else so moved back to `command.cpp`

klausler commented on flang/runtime/tools.h:
```diff
@@ -411,6 +412,48 @@ RT_API_ATTRS void ShallowCopy(const Descriptor &to, const Descriptor &from,
+RT_API_ATTRS void CheckAndStoreIntToDescriptor(
+    const Descriptor *intVal, std::int64_t value, Terminator &terminator);
+
+template <int KIND> struct FitsInIntegerKind;
```
What does this do?

yi-wu-arm replied:
check if it can hold a int64, moved back to `command.cpp` since it is only used there.

klausler commented on flang/runtime/tools.cpp:
```diff
@@ -173,5 +173,141 @@ RT_API_ATTRS void ShallowCopy(const Descriptor &to, const Descriptor &from) {
+  return static_cast<std::int64_t>(std::strlen(string));
+}
+
+// Assumes Descriptor \p value is not nullptr.
```
The comment is incorrect; the code works fine if `value` is null.

yi-wu-arm replied:
right, changed the comment and content of `IsValidIntDecsiptor` as well, so both of them can work with `null`.

klausler commented on flang/runtime/tools.cpp:
```diff
@@ -173,5 +173,141 @@ RT_API_ATTRS void ShallowCopy(const Descriptor &to, const Descriptor &from) {
+}
+
+// Assumes Descriptor \p intVal is not nullptr.
+RT_API_ATTRS bool IsValidIntDescriptor(const Descriptor *intVal) {
```
This function could easily accommodate a null `value`.

yi-wu-arm replied:
correct, comment removed

klausler commented on flang/runtime/tools.cpp:
```diff
@@ -173,5 +173,141 @@ RT_API_ATTRS void ShallowCopy(const Descriptor &to, const Descriptor &from) {
+}
+
+// Assume Descriptor \p value is valid: pass IsValidCharDescriptor check.
+RT_API_ATTRS void FillWithSpaces(const Descriptor &value, std::size_t offset) {
```
What if `offset` is larger than `ElementBytes()`?

yi-wu-arm replied:
there is an if check, compare with `ElementBytes()`, moved back to `command.cpp` since it is only used there.

yiwu0b11 committed:
Revert "move StoreToDescriptor functions to tools.h"

This reverts commit 438362d74d7ac9cc74645714afc20ab9b3616385.

yiwu0b11 committed:
small fixes

yiwu0b11 committed:
clang-format

klausler commented on flang/include/flang/Runtime/execute.h:
```diff
@@ -0,0 +1,31 @@
+
+#include "flang/Runtime/entry-names.h"
+
+#include <cstdint>
```
Is this `#include` necessary to define anything in this header?

yi-wu-arm replied:
no, removed.

yi-wu-arm replied:
not anymore, should be removed

klausler commented on flang/runtime/tools.h:
```diff
@@ -411,6 +412,24 @@ RT_API_ATTRS void ShallowCopy(const Descriptor &to, const Descriptor &from,
+// If descriptor `errmsg` is provided, error messages will be stored to it.
+// Returns stats specified in standard.
+RT_API_ATTRS std::int32_t CopyCharsToDescriptor(const Descriptor &value,
+    const char *rawValue, std::int64_t rawValueLength,
```
`rawValueLength` should probably be `std::size_t`.

yi-wu-arm replied:
done

klausler commented on flang/runtime/tools.cpp:
```diff
@@ -173,5 +173,70 @@ RT_API_ATTRS void ShallowCopy(const Descriptor &to, const Descriptor &from) {
 }
 
+RT_API_ATTRS const char *EnsureNullTerminated(
+    const char *str, size_t length, Terminator &terminator) {
```
`std::size_t`

yi-wu-arm replied:
done

klausler commented on flang/runtime/tools.cpp:
```diff
@@ -173,5 +173,70 @@ RT_API_ATTRS void ShallowCopy(const Descriptor &to, const Descriptor &from) {
 
+RT_API_ATTRS const char *EnsureNullTerminated(
+    const char *str, size_t length, Terminator &terminator) {
+  if (length <= std::strlen(str)) {
```
Do not call `std::strlen()` on a string that is not known to be NUL terminated.  Use `std::memchr` instead.

yi-wu-arm replied:
done, thank you for taking the time to review my code, especially during the holiday season!

klausler commented on flang/runtime/execute.cpp:
```diff
@@ -0,0 +1,203 @@
+
+    // Convert the char to wide char
+    const size_t sizeNeeded{mbstowcs(NULL, newCmdWin, 0) + 1};
+    wchar_t *wcmd{new wchar_t[sizeNeeded]};
```
Do not use `operator new` here.  Use `AllocateMemoryOrCrash()` like elsewhere.

yi-wu-arm replied:
used `AllocateMemoryOrCrash()` to allocate memory

klausler commented on flang/runtime/execute.cpp:
```diff
@@ -0,0 +1,203 @@
+    }
+#endif
+  }
+  FreeMemory((void *)newCmd);
```
Do not call `FreeMemory` on `newCmd` unless `newCmd` was allocated in `EnsureNullTerminated`.

yi-wu-arm replied:
added an if condition `newCmd != command.OffsetElement()`

yiwu0b11 committed:
memory fixes and use std::size_t

yiwu0b11 committed:
change std::int64 to std::size_t

yiwu0b11 committed:
windows fixes

klausler commented on flang/runtime/execute.cpp:
```diff
@@ -0,0 +1,206 @@
+    si.cb = sizeof(si);
+    ZeroMemory(&pi, sizeof(pi));
+
+    // append "cmd.exe /c " to the beginning of command
```
"append" means adding on to the end of something.

yi-wu-arm replied:
replaced by `add`

klausler commented on flang/runtime/execute.cpp:
```diff
@@ -0,0 +1,206 @@
+    }
+#endif
+  }
+  // Deallocate memory if EnsureNullTerminated dynamically allocate a memory
```
"allocated memory"

klausler commented on flang/runtime/tools.cpp:
```diff
@@ -173,5 +173,72 @@ RT_API_ATTRS void ShallowCopy(const Descriptor &to, const Descriptor &from) {
+    const char *str, std::size_t length, Terminator &terminator) {
+  const void *nullTerminatorPos{std::memchr(str, '\0', length)};
+
+  if (nullTerminatorPos == nullptr) {
```
`if (std::memchr(str, '\0', length) == nullptr) {`

yi-wu-arm replied:
done

yiwu0b11 committed:
grammar and minor fixes

yi-wu-arm commented:
Hi, @jeanPerier welcome back! I've made quite a few changes, mainly on memory allocate and deallocate, test improvements, and have moved some commonly used descriptor functions into `tools.cpp` to re-use some code instead of creating duplicates of them. Could you review the recent changes? Thanks in advance!

jeanPerier reviewed (approved):
Thanks for the updates @yi-wu-arm, LGTM

yi-wu-arm commented on flang/runtime/execute.cpp:
```diff
@@ -0,0 +1,206 @@
+    FreeMemory((void *)wcmd);
+#else
+    // terminated children do not become zombies
+    signal(SIGCHLD, SIG_IGN);
```
This actually cause a crash on async execution. Once an async `execute_command_line` was called, all `execute_command_line` after that will result in `std::system` returns a status code of -1. 
Comment out this line would be a temporary fix.

yi-wu-arm replied:
reproducer: aarch64, linux, clang++
```cpp
#include <unistd.h>
#include <signal.h>
#include <iostream>
int execute_command_line(const char *cmd, bool isWait){
    if(isWait){
        return std::system(cmd);
    }else{
        signal(SIGCHLD, SIG_IGN);
        pid_t pid=fork();
        if(pid<0){
            return -999;
        }else if(pid==0){
            return std::system(cmd);
        }
        exit(pid);
        return -111;
    }
}

int main() {
    std::cout << execute_command_line("InvalidCommand", false) << std::endl;
    std::cout << execute_command_line("InvalidCommand", true) << std::endl;

    return 0;
}
```
console printout:
```
sh: 1: InvalidCommand: not found
-1
sh: 1: InvalidCommand: not found
-1
```
console printout when `signal(SIGCHLD, SIG_IGN);` is commented out.
```
sh: 1: InvalidCommand: not found
32512
sh: 1: InvalidCommand: not found
32512
```

yi-wu-arm replied:
or an alternativelydo what `gfortran` is doing: do not change the value of `exitstatus`, `cmdmsg`, and set `cmdstatus` to 0, no mater what happened.

```fortran
program test_associated
  integer :: exitstatus, cmdstatus
  character(len=35) :: cmdmsg
  exitstatus=404
  cmdstatus=303
  cmdmsg="Does async execution change cmdmsg?"

  call execute_command_line ("InvalidCommand", .false., exitstatus, cmdstatus, cmdmsg)
  call sleep(1)

  print *, "Exist status was: ", exitstatus
  print *, "Command status was: ", cmdstatus
  print *, "Error message is: ", trim(cmdmsg)

  end
  ```
  console output when `isWait=false` (async)
  ```
  sh: 1: InvalidCommand: not found
 Exist status was:          404
 Command status was:            0
 Error message is: Does async execution change cmdmsg?
  ```
  
  console output when `isWait=true` (sync)
  ```
  sh: 1: InvalidCommand: not found
 Exist status was:          127
 Command status was:            3
 Error message is: Invalid command line
 ```
 
 Note that `call sleep()` is not available on flang-new, so on llvm-test-suite https://github.com/llvm/llvm-test-suite/blob/9ca97f5027150f7e507e5ab4c56f38a29fb3c696/Fortran/gfortran/regression/execute_command_line_1.f90#L16, it would fail to compile as well,

yi-wu-arm replied:
It turns out that resetting the signal doesn't work well. put `signal(SIGCHLD, SIG_IGN)` will still cause system return -1. However, using a non-blocking `waitpid` to get the exit code of child process will clear up the zombie. (I didn't there is an non-blocking wait).
```cpp
void sigchld_handler(int signo) {
    int status;
    waitpid(-1, &status, WNOHANG); // WNOHANG means non-blocking
}
.......
signal(SIGCHLD, sigchld_handler);
pid_t pid{fork()};
if (pid < 0) {
  error;
} else if (pid == 0) {
  std::system(cmd);
}
```

tblah replied:
Sounds good to me :+1: 

Out of interest, what did you try to reset the handler for SIGCHLD?

yi-wu-arm replied:
Opps, sorry I type it wrong in the comment, I mean `signal(SIGCHLD, SIG_DFL)`

Comment updated.

dankm commented on flang/runtime/execute.cpp:
```diff
@@ -0,0 +1,206 @@
+  int exitStatusVal{status};
+  if (exitStatusVal == 1) {
+#else
+  int exitStatusVal{WEXITSTATUS(status)};
```
Linux defines WEXITSTATUS in stdlib.h (and therefore cstdlib), but at least FreeBSD and NetBSD require including sys/wait.h.

It should be harmless to always include sys/wait.h, so I'd suggest doing that.

dankm replied:
Created #77675 for this.

psteinfeld commented:
See issue #77975.

psteinfeld commented:
See issue #77984

psteinfeld commented:
See issue #77979.
